%\VignetteIndexEntry{intcomp}
%The above line is needed to remove a warning in R CMD check
\documentclass[a4paper]{article}

\title{intcomp:\\Benchmarking tool for gene expression - copy number data integration models}

\author{Leo Lahti\footnote{leo.lahti@iki.fi} and Martin Sch{\"a}fer}

\usepackage{Sweave}
\usepackage{float}

\usepackage{amsmath,amssymb,amsfonts}
%\usepackage[authoryear,round]{natbib}


\usepackage{hyperref}
\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}

\begin{document}

\maketitle

\section{Introduction}

Several algorithms have been suggested to integrate gene expression
and DNA copy number measurement to discover novel cancer-associated
genes and chromosomal regions. However, quantitative comparison of
these models has been missing. This R package provides a benchmarking
pipeline for integrative cancer gene detection methods. Cancer gene
detection performance of each algorithm is evaluated by comparing the
prioritized candidate gene list from each method to a golden standard
list of known cancer genes in simulated and real data sets.

The package is experimental beta-release and provided as is. The main
purpose of this vignette is to provide reference to algorithmic
details of the benchmarking framework used in \cite{Lahti11review} and
sufficient documentation of the functionality such that similar
experiments can be replicated by experienced R users.


\subsection{Comparison methods}

The current version implements the comparison of the following ge/cn
 integration algorithms: CNAmet \cite{Hautaniemi04,Louhimo11}, DRI
 \cite{Salari2010}, edira \cite{Schafer09}, intCNGEan
 \cite{Wieringen09}, Ortiz-Estevez \cite{Ortiz-Estevez11}, pint
 \cite{Lahti09mlsp}, PMA \cite{Witten09}, PREDA/SODEGIR
 \cite{Bicciato09}, and SIM \cite{Menezes2009}. Multiple variants are
 evaluated for some methods \cite{Lahti11review}.


\section{Examples}


\subsection{Installing dependencies}

The benchmarking pipeline depends on various external R
packages. Install the  dependencies from within R using the following
commands:

<<deps, results = hide>>=
source("http://www.bioconductor.org/biocLite.R") 
biocLite(c("biomaRt", "DNAcopy", "DRI", "edira", 
"hgu133ahsentrezg.db", "intCNGEan", "org.Hs.eg.db", 
"PMA", "SIM", "PREDA", "CGHcall", "CNAmet"))
@

You may need to install the following packages manually: curl
library\footnote{http://curl.haxx.se/download.html} and the R packages
XML\footnote{http://cran.r-project.org/web/packages/XML/index.html},
RCurl\footnote{http://www.omegahat.org/RCurl/},
edira\footnote{http://www.statistik.tu-dortmund.de/~schaefer/},
intCNGEan\footnote{http://www.few.vu.nl/~wvanwie/software/intCNGEan/intCNGEan.html},
org.Hs.eg.db\footnote{http://www.bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html},
PREDA/SODEGIR\footnote{http://www.xlab.unimo.it/PREDA/PREDAinstall.R;
http://www.xlab.unimo.it/PREDA/PREDAsampledata\_0.1.7.tar.gz;
http://www.xlab.unimo.it/PREDA/PREDA\_0.2.12.tar.gz},
and CNAmet\footnote{http://csbi.ltdk.helsinki.fi/CNAmet/}.

\section{Running the comparison tests}

\subsection{Loading example data}

The package contains the data sets from \cite{Hyman2002} and
\cite{Pollack2002}. These two data sets are available from public
sources and have been used to validate multiple alternative
integration algorithms.

\subsubsection{Pollack et al. (2002)}

Loading Pollack et al. (2002) \cite{Pollack2002} data
set\footnote{http://www.pnas.org/content/suppl/2002/09/23/162471999.DC1/4719CopyNoGeneDatsetLegend.html
  accessed June 2, 2010.}.

<<results=tex>>=
library(intcomp)
data(pollack)
dat <- read.pollack(chrs = 1:22, CopyNoGeneDataset4719, clone2geneid)
ge <- dat$ge$data        # gene expression matrix
cn.raw <- dat$cn$data        # copy number matrix
gene.info <- dat$ge$info # gene location info
@

\subsubsection{Hyman et al. (2002)}

Loading Hyman et al. (2002) \cite{Hyman2002} data
set\footnote{HymancdnaDataA.tab, HymancghDataA.tab and HymanAcc.mat
  obtained from http://www.ece.ucsb.edu/pubs/ieee/index.shtml accessed
  June 2, 2010.}.

<<results=tex>>=
data(hyman)
library("org.Hs.eg.db")
dat <- read.hyman(cdna, cgh, genenames, chrs = 1:22, as.list(org.Hs.egALIAS2EG))
ge <- dat$ge        # gene expression matrix + info
cn.raw <- dat$cn    # copy number matrix + info
gene.info <- dat$cn.raw$info # gene location info
@

\subsection{Preprocessing copy number data}

Some methods require segmented and/or called copy number data, obtained with

<<preprocess.cn, results = hide>>=
cgh <- process.copynumber(cn.raw)
cn.seg <- list(data = assayDataElement(cgh, 'segmented'), info = gene.info)
cn.call <- list(data = assayDataElement(cgh, 'calls'), info = gene.info)
rownames(cn.call$data) <- rownames(cn.seg$data) <- rownames(ge$data)
cn.call$info <- cn.seg$info <- cn.raw$info
@ 


\subsection{Golden standard list of known cancer genes}

The golden standard list of known breast cancer genes from The Breast
Cancer Gene
Database\footnote{http://www.tumor-gene.org/cgi-bin/TGDB/tgdb\_by\_name.cgi
accessed 5.6.2010; 'tgdb\_by\_name.cgi.html' and 'tgdb.txt'}
\cite{Baasiri99} was downloaded and stored to the tgdb object. This
script details how to load the list, convert to Entrez GeneIDs and
select the genes that are included in the experimental data:

<<results=tex>>=
library("org.Hs.eg.db")
data(tgdb)
cancerGenes <- get.brca.genes(rownames(dat$ge$data), as.list(org.Hs.egALIAS2EG), tgdb)
@

\subsection{Simulating data}

To evaluate the methods based on simulations, data can be simulated following roughly the approach given in \cite{Schafer09}, but with greater flexibility. The quantile grid to be simulated can be defined
entirely by the user, as well as the mixing weight, the number of different variances to be considered and the call probabilities (for testing {\it intcngean}).

<<test_schaefer, results = hide, eval=FALSE>>=
library(ediraAMLdata)
data(AMLdata, package="ediraAMLdata")
test_schaefer <- test.simulation(GE, CN, probespanGE = 16, probespanCN = 100,
method="schaefer", Inner=3:5, Outer=c(1:2,6:7),
probs_GE=c(0.025,0.075,0.3,0.5,0.7,0.925,0.975),
probs_CN=c(0.025,0.075,0.3,0.5,0.7,0.925,0.975),
cancer_GE=c(1,1,2,2,6,6,7,7),
cancer_CN=c(1,2,1,2,3,4,3,4),
n=100, weight=1/10, variances=c(1/4,1/2,1,2,4),
GE_norm=4, CN_norm=2, seed=42,
call_probs=c(0.001,0.005,0.01,0.0125,0.04,0.925,0.99))
@

Also, testing on the data simulated by Francesco Ferrari is provided. Here, the data are given, i.e., the user has no influence on their characteristics.
<<test_ferrari, results = hide, eval=FALSE>>=
test_ferrari <- test.simulation(GE, CN, probespanGE = 16, probespanCN = 100,
method="ferrari")
@

\section{Running the benchmarking pipeline}

The example data set now contains (i) gene expression data (ge), (ii)
gene copy number data (cn), (iii) optional sample class labels
(tumor/normal), and (iv) golden standard list of known cancer
genes. The ge and cn data sets are lists containing $data$ and $info$
fields. The probes in gene expression and gene copy number are paired;
$data$ is a data matrix with gene expression (ge\$data) or gene copy
number (cn\$data) data; $info$ field is a data frame containing
additional information about genes: $loc$ indicates the genomic
location of the probes in base pairs (numeric); $chr$ and $arm$ are
factors indicating the chromosome and chromosomal arm of the probe,
respectively. To run the benchmarking tests for all methods, use:

<<run, results = hide, eval=FALSE>>=
methods <- c("CNAmet","edira")
res_real <- test.geneorder.pipeline(ge = ge, cn.raw = cn.raw,
cn.seg = cn.seg, cn.call = cn.call, cghCall = cgh,
Labels = NULL, cancerGenes = cancerGenes, nperm=20,
input = "real", version = "normal",
methods = methods, references="none")
auc.ordered <- sort(unlist(res_real$auc))
@

<<run.simulation, results = hide, eval=FALSE>>=
methods <- c("CNAmet","edira")
res_simulated  <- test.geneorder.pipeline(ge=test_schaefer$ge,
cn.raw=test_schaefer$cn.raw, cn.seg=NULL,
cn.call=test_schaefer$cn.call,
cghCall = test_schaefer$cn.cghCall,
ge.norm = test_schaefer$ge.norm,
cn.norm = test_schaefer$cn.norm,
Labels=test_schaefer$Labels,
cancerGenes=test_schaefer$cancerGenes,
nperm = 20, input="simulations.equal.dimensions",
version = "normal",
methods=methods, callprobs=sim$callprobs,
references="both")
auc.ordered <- sort(unlist(res_simulated$auc))
@

If segmented data is supplied, then it will be used for analysis.
If segmented data is missing, then raw data will be used if supplied.
If both segmented and raw data are missing, then called data will be used if supplied.\\

The following {\it methods} are available: {\it OrtizEstevez}, {\it
  intcngean}, {\it edira}, {\it pint}, {\it SIM.window\}, {\it
    SIM.full\}, {\it DRI.cp}, {\it DRI.cs}, {\it PMA.raw}, {\it
      CNAmet}, {\it PREDA} and for two-group comparisons (requiring
    Labels) {\it DRI.ct}, {\it DRI.ss}, {\it DRI.srank}, {\it
      DRI.sraw}. The cancer gene prioritization of each method is
    compared to the golden standard list of known cancer genes; the
    result contains running times of the algorithms and the AUC values
    from ROC analysis \cite{Lahti11review}. The AUC values provide
    quantitative estimates of model performance in cancer gene
    detection and provide the basis for comparisons.

\subsection*{Acknowledgements}

The package has been supported by EuGESMA COST Action BM0801: European
Genetic and Epigenetic Study on AML and MDS.

%* references

\bibliographystyle{ieeetr}
\bibliography{intcomp}




\end{document}
