%\VignetteIndexEntry{intcomp}
%The above line is needed to remove a warning in R CMD check
\documentclass[a4paper]{article}

\title{intcomp:\\Benchmarking pipeline for integrative cancer gene
prioritization algorithms based on gene expression and copy number data}

\author{Leo Lahti\footnote{leo.lahti@iki.fi} and Martin Sch{\"a}fer}

\usepackage{Sweave}
\usepackage{float}

\usepackage{amsmath,amssymb,amsfonts}
%\usepackage[authoryear,round]{natbib}


\usepackage{hyperref}
\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}

\begin{document}

\maketitle

\section{Introduction}

Several algorithms have been suggested to integrate gene expression
and DNA copy number measurement to discover cancer-associated
chromosomal regions, but quantitative comparison of these models has
been missing. The \Rpackage{intcomp} R package provides a benchmarking
pipeline for quantitative comparisons between the different
implementations for integrative analysis of ge/cn data. This vignette
provides installation instructions, practical examples and references
to the algorithmic details of the \Rpackage{intcomp} benchmarking
pipeline \cite{Lahti11review}.

In the \Rpackage{intcomp} pipeline, the cancer gene detection
performance of each algorithm is evaluated based on gene
prioritization by using each method. Each method is used to order the
gene list, and the resulting order is compared to golden standard
lists of known cancer genes on simulated and real data sets. For
details, see \cite{Lahti11review}.

\section{Installation}

\subsection{Installing the intcomp benchmarking pipeline}

To install this package directly within R type:

<<inst, results = hide, eval = FALSE>>=
install.packages("intcomp", type = "source", 
  repos="http://R-Forge.R-project.org", dependencies = TRUE)
@

In case of error messages, see below.

\subsection{Dependencies}

You may need to install dependencies before the intcomp package can be
installed. The benchmarking pipeline depends on various external R
packages. Install the dependencies from within R using the following
commands:

<<deps, results = hide, eval = FALSE>>=
source("http://www.bioconductor.org/biocLite.R") 
biocLite(c("biomaRt", "DNAcopy", "DRI", "edira", 
"hgu133ahsentrezg.db", "intCNGEan", "org.Hs.eg.db", 
"PMA", "SIM", "PREDA", "CGHcall", "CNAmet"))
@

You may need to install the following packages manually: curl
library\footnote{http://curl.haxx.se/download.html} and the R packages
XML\footnote{http://cran.r-project.org/web/packages/XML/index.html},
RCurl\footnote{http://www.omegahat.org/RCurl/},
edira\footnote{http://www.statistik.tu-dortmund.de/~schaefer/},
intCNGEan\footnote{http://www.few.vu.nl/~wvanwie/software/intCNGEan/intCNGEan.html},
org.Hs.eg.db\footnote{http://www.bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html},
PREDA/SODEGIR\footnote{http://www.bioconductor.org/packages/devel/bioc/html/PREDA.html},
and CNAmet\footnote{http://csbi.ltdk.helsinki.fi/CNAmet/}.

\section{Benchmarking the comparison methods}

The package contains a copy of the publicly available cancer data sets
from \cite{Hyman2002} and \cite{Pollack2002} to benchmark the cancer
gene detection algorithms on real experimental data, and two simulated
data sets from previous publications \cite{Bicciato09, Schafer09},
called Ferrari and Schaefer data sets, respectively. The data sets
have varying setups, depending on whether they include two-group
comparisons or segmented/called copy number data. Showcases running
the benchmarking pipeline on each data set are described below.

\subsection{Hyman et al. (2002)}

The Hyman et al. (2002) \cite{Hyman2002} breast cancer data
set\footnote{HymancdnaDataA.tab, HymancghDataA.tab and HymanAcc.mat
obtained from http://www.ece.ucsb.edu/pubs/ieee/index.shtml accessed
June 2, 2010.}, and a golden standard list of known breast cancer
genes from The Breast Cancer Gene Database \cite{Baasiri99} provide
the first example data set for benchmarking the comparison algorithms.
The cancer gene list was
downloaded\footnote{http://www.tumor-gene.org/cgi-bin/TGDB/tgdb\_by\_name.cgi
accessed 5.6.2010; 'tgdb\_by\_name.cgi.html' and 'tgdb.txt'} and
stored to the tgdb object. The gene symbols are converted into Entrez
Gene IDs, the probes are matched between gene expression and copy
number data, as detailed in the {\it read.hyman} function, and the
known breast cancer genes from the TGDB golden standard list present
in the ge/cn data are selected. Further details are detailed in the
{\it read.hyman} and {\it the get.brca.genes} functions. 

For Hyman, the original non-segmented data set from the publication is
used (cn.seg = cn.raw) in the experiments (except with intCNGEan and
CNAmet that require segmented and called data, respectively). To run
the intcomp benchmarking pipeline on Hyman data set, use

<<hyman, results=hide, eval = FALSE>>=
# Set comparison methods for benchmarking
methods <- c("CNAmet","edira")

# Load the benchmarking pipeline
library(intcomp)

# Load and preprocess the Hyman data
data(hyman)
library("org.Hs.eg.db")
symbol2entrezid <- as.list(org.Hs.egALIAS2EG)
hyman <- read.hyman(cdna, cgh, genenames, xx = symbol2entrezid)

# Load the known breast cancer genes
data(tgdb)
# Convert to EntrezIDs and select genes common with the ge/cn data
cancerGenes <- get.brca.genes(rownames(hyman$ge$data), symbol2entrezid, tgdb)

# Run the benchmarking pipeline
res.hyman <- test.geneorder.pipeline(ge = hyman$ge,           # expression
                                     cn.raw = hyman$cn.raw,   # raw copy number
				     cghCall = hyman$cghCall, # segm/called cn
				     cancerGenes = cancerGenes, 
				     methods = methods, 
				     cn.default = "raw", # use with ge 
             references="none")  # no real reference data included 

# Check the AUC values for the methods
auc.ordered <- sort(unlist(res.hyman$auc))
@

\subsection{Pollack et al. (2002)}

The Pollack et al. (2002) \cite{Pollack2002} data
set\footnote{http://www.pnas.org/content/suppl/2002/09/23/162471999.DC1/4719CopyNoGeneDatsetLegend.html
accessed June 2, 2010.} is also used in combination with the golden
standard list from the TGDB (See Hyman data set). The gene identifiers
in the Pollack data are converted into Entrez Gene IDs. To run the
benchmarking tests on Pollack data set, use

<<pollack, results = hide, eval = FALSE>>=
# Set comparison methods for benchmarking
methods <- c("CNAmet","edira")

# Load and preprocess the GE/CN data
library(intcomp)
data(pollack)
pollack <- read.pollack(dat = CopyNoGeneDataset4719, clone2geneid = clone2geneid)

# Load the known breast cancer genes
library("org.Hs.eg.db")
data(tgdb)
# Convert to EntrezIDs and select genes common with the ge/cn data
cancerGenes <- get.brca.genes(rownames(pollack$ge$data), as.list(org.Hs.egALIAS2EG), tgdb)

# Run the benchmarking pipeline
res.pollack <- test.geneorder.pipeline(ge = pollack$ge, 
                                       cn.raw = pollack$cn.raw,
				       cghCall = pollack$cghCall,
				       cancerGenes = cancerGenes, 
				       methods = methods, 
				       cn.default = "raw",
               references="none") 

# Check the AUC values for the methods
auc.ordered <- sort(unlist(res.pollack$auc))

@

\subsection{Ferrari data set (2009)}

The first simulated data set, where the exact ground truth is known,
is provided by the simulation approach given in \cite{Bicciato09}:

<<ferrari, results = hide, eval=FALSE>>=
# Generate simulated data
library(intcomp)
ferrari <- test.simulation(GE, CN, method = "ferrari")

# Run the benchmarking pipeline
res.ferrari  <- test.geneorder.pipeline(
       ge = ferrari$ge, 
  cn.raw  = ferrari$cn.raw,
  cn.seg  = ferrari$cn.seg, 
  cn.call  = ferrari$cn.call,
  cghCall = ferrari$cn.cghCall, 
  cancerGenes = ferrari$cancerGenes,
      methods = methods)

# Check the AUC values from comparison methods
auc.ordered <- sort(unlist(res.ferrari$auc))
@

\subsection{Schaefer data set (2009)}

The second simulated data set is provided by the simulation approach
given in \cite{Schafer09} with added flexibility. The quantile grid to
be simulated can be defined by the user, as well as the mixing weight,
the number of different variances to be considered and the call
probabilities.

<<schaefer, results = hide, eval=FALSE>>=

# Set the methods for benchmarking
methods <- c("CNAmet","edira")

# Generate simulated data
library(intcomp)
library(ediraAMLdata)
data(AMLdata, package = "ediraAMLdata")
schaefer <- test.simulation(GE, CN, method = "schaefer")

# Run the benchmarking pipeline
res.schaefer  <- test.geneorder.pipeline(
       ge = schaefer$ge,
   cn.raw = schaefer$cn.raw, 
  cghCall = schaefer$cn.cghCall,
 cancerGenes = schaefer$cancerGenes,
  methods = methods, 
  callprobs = schaefer$callprobs,
  cn.default = "raw")
  
# Check the AUC values from comparison methods
auc.ordered <- sort(unlist(res.schaefer$auc))
@


\section{Notes on the benchmarking pipeline}

The minimal input data for the test.geneorder.pipeline banchmarking
function includes (i) gene expression data (ge), (ii) gene copy number
data (cn.raw / cn.seg / cn.call / cghCall), (iii) a golden standard
list of known cancer genes (cancerGenes), and (iv) the list of methods
to compare (mehods).

The gene expression and copy number data sets are lists containing
$data$ and $info$ fields; the probes in gene expression and gene copy
number need to be matched; $data$ is a data matrix with gene
expression (GE\$data) or gene copy number (CN\$data) data; $info$
field is a data frame containing additional information about genes:
$loc$ indicates the genomic location of the probes in base pairs
(numeric); $chr$ and $arm$ are factors indicating the chromosome and
chromosomal arm of the probe, respectively. The user can provide the
copy number data as raw (cn.raw), segmented (cn.seg) or called
(cn.call) version. Certain methods require specific versions of the
copy number. For instance, the CNAmet requires called copy number
data. The intCNGEan algorithm requires copy number as a cghCall object
from the \Rpackage{CGHcall} R package. It is advisable to provide all
four versions - cn.raw, cn.seg, cn.call and cghCall - in the input to
the test.geneorder.pipeline function when possible. The cn.raw, cn.seg
and cn.call should follow the data + info format explained above, and
the cghCall contains the raw, segmented and called data in the cghCall
format. Finally, if multiple versions of copy number data are
available, the user can specify (through the cn.default argument)
which version is coupled with gene expression data unless otherwise
specified by particular methods. By default, the associations between
gene expression and segmented copy number data (ge + cn.seg) are
investigated.


\section{Comparison methods}

The following {\it implementations} are available in the
\Rpackage{intcomp} benchmarking pipeline: {\it CNAmet}
\cite{Hautaniemi04,Louhimo11}, variants of {\it DRI}
\cite{Salari2010}, {\it edira} \cite{Schafer09}, {\it intcngean}
\cite{Wieringen09}, {\it OrtizEstevez} \cite{Ortiz-Estevez11}, {\it
pint} \cite{Lahti09mlsp}, variants of {\it SIM} \cite{Menezes2009}, 
{\it PMA} \cite{Witten09}, {\it PREDA/SODEGIR}
\cite{Bicciato09,Ferrari11}. The list of available methods in the
pipeline is retrieved with:
    
<<methods,results=hide>>=
library(intcomp)
@

<<methods2>>=
list.methods()
@


\section{Benchmarking results}
      
The prioritized cancer gene list provided by each method is compared
to the golden standard list of known cancer genes; the result contains
running times of the algorithms and the AUC values from ROC analysis.
The AUC values provide quantitative estimates of model performance in
cancer gene detection and provide the basis for the comparisons as
reported in \cite{Lahti11review}.

\subsection{Version details}

The following package versions were used to produce this vignette:

<<sessioninfo>>=
sessionInfo()
@

\subsection*{Acknowledgements}

This work has been supported by EuGESMA COST Action BM0801: European
Genetic and Epigenetic Study on AML and MDS. We would also like to
thank Fransesco Ferrari for providing simulated data for the study.

%\bibliographystyle{ieeetr}
%\bibliography{intcomp}

\begin{thebibliography}{10}

\bibitem{Lahti11review}
L.~Lahti, M.~Sch{\"a}fer, H.-U. Klein, S.~Bicciato, and M.~Dugas, ``Cancer gene
  prioritization by multi-platform data integration: a comparative review.
  upcoming,''

\bibitem{Hautaniemi04}
S.~Hautaniemi, M.~Ringn{\'e}r, P.~Kauraniemi, R.~Autio, H.~Edgren,
  O.~Yli-Harja, J.~Astola, A.~Kallioniemi, and O.~Kallioniemi, ``A strategy for
  identifying putative causes of gene expression variation in human cancers,''
  {\em Journal of the Franklin Institute}, vol.~341, no.~1-2, pp.~77--88, 2004.

\bibitem{Louhimo11}
R.~Louhimo and S.~Hautaniemi, ``{CNAmet: an R package for integrating copy
  number, methylation and expression data},'' {\em Bioinformatics}, vol.~27,
  no.~6, pp.~887--888, 2011.

\bibitem{Salari2010}
K.~Salari, R.~Tibshirani, and J.~R. Pollack, ``{DR-Integrator: a new analytic
  tool for integrating DNA copy number and gene expression data.},'' {\em
  Bioinformatics}, vol.~26, no.~3, pp.~414--6, 2010.

\bibitem{Schafer09}
M.~Sch{\"a}fer, H.~Schwender, S.~Merk, C.~Haferlach, K.~Ickstadt, and M.~Dugas,
  ``{Integrated analysis of copy number alterations and gene expression: a
  bivariate assessment of equally directed abnormalities},'' {\em
  Bioinformatics}, vol.~25, no.~24, pp.~3228--3235, 2009.

\bibitem{Wieringen09}
W.~N. van Wieringen and M.~A. van~de Wiel, ``{Nonparametric testing for DNA
  copy number induced differential mRNA gene expression},'' {\em Biometrics},
  vol.~65, pp.~19--29, 2009.

\bibitem{Ortiz-Estevez11}
M.~Ortiz-Estevez, J.~De~Las~Rivas, C.~Fontanillo, and A.~Rubio, ``Segmentation
  of genomic and transcriptomic microarrays data reveals major correlation
  between {DNA} copy number aberrations and gene-loci expression,'' {\em
  Genomics}, vol.~97, pp.~86--93, 2011.

\bibitem{Lahti09mlsp}
L.~Lahti, S.~Myllykangas, S.~Knuutila, and S.~Kaski, ``Dependency detection
  with similarity constraints,'' in {\em Proceedings MLSP'09 IEEE International
  Workshop on Machine Learning for Signal Processing XIX}, (Piscataway, NJ),
  pp.~89--94, IEEE Signal Processing Society, September 2-4 2009.
\newblock Implementation available in pint package of R/BioConductor
  http://www.bioconductor.org/packages/release/bioc/html/pint.html.

\bibitem{Witten09}
D.~M. Witten, R.~Tibshirani, and T.~Hastie, ``{A penalized matrix
  decomposition, with applications to sparse principal components and canonical
  correlation analysis},'' {\em Biostatistics}, vol.~10, no.~3, pp.~515--534,
  2009.

\bibitem{Bicciato09}
S.~Bicciato, R.~Spinelli, M.~Zampieri, E.~Mangano, F.~Ferrari, L.~Beltrame,
  I.~Cifola, C.~Peano, A.~Solari, and C.~Battaglia, ``{A computational
  procedure to identify significant overlap of differentially expressed and
  genomic imbalanced regions in cancer datasets},'' {\em Nucleic Acids
  Research}, vol.~37, pp.~5057--5070, 2009.

\bibitem{Menezes2009}
R.~X. Menezes, M.~Boetzer, M.~Sieswerda, G.-J.~B. van Ommen, and J.~M. Boer,
  ``{Integrated analysis of DNA copy number and gene expression microarray data
  using gene sets.},'' {\em BMC bioinformatics}, vol.~10, no.~1, p.~203, 2009.

\bibitem{Hyman2002}
E.~Hyman, P.~Kauraniemi, S.~Hautaniemi, M.~Wolf, S.~Mousses, E.~Rozenblum,
  M.~Ringner, G.~Sauter, O.~Monni, A.~Elkahloun, O.-P. Kallioniemi, and
  A.~Kallioniemi, ``{Impact of DNA Amplification on Gene Expression Patterns in
  Breast Cancer},'' {\em Cancer Res.}, vol.~62, pp.~6240--6245, Nov. 2002.

\bibitem{Pollack2002}
J.~R. Pollack, T.~S\o~rlie, C.~M. Perou, C.~A. Rees, S.~S. Jeffrey, P.~E.
  Lonning, R.~Tibshirani, D.~Botstein, A.-L. B\o~rresen Dale, and P.~O. Brown,
  ``{Microarray analysis reveals a major direct role of DNA copy number
  alteration in the transcriptional program of human breast tumors.},'' {\em
  Proceedings of the National Academy of Sciences of the United States of
  America}, vol.~99, pp.~12963--8, Oct. 2002.

\bibitem{Baasiri99}
R.~Baasiri, S.~Glasser, D.~Steffen, and D.~Wheeler, ``{The Breast Cancer Gene
  Database: a collaborative information resource},'' {\em Oncogene}, vol.~18,
  no.~56, pp.~7958--7965, 1999.

\bibitem{Ferrari11}
Ferrari F, Solari A, Battaglia C, Bicciato S. PREDA: an R-package to
identify regional variations in genomic data. {\em Bioinformatics}
online July 7, 2011. 


\end{thebibliography}


\end{document}
